### PR TITLE
New model manageiq manageiq user

### PR DESCRIPTION
**Pull request by yaacov**
**SUMMARY**

ManageIQ is an open source management platform for Hybrid IT.

This change is adding:

    manageiq_user module, responsible for user management in ManageIQ
    manageiq utils
    manageiq doc_fragment

All the modules, including docs, tests and usage examples can be found here

Currently, the only requirement for the modules is manageiq-api-client-python

**ISSUE TYPE**

    New Module Pull Request

**COMPONENT NAME**

manageiq_user.py (module)
manageiq.py (module utils and doc_fragement)

**ANSIBLE VERSION**

ansible 2.3.0.0

### CODE REVIEW

**Comment by yaacov** [comment]
@dkorn @gundalow @thaumos @cben @simon3z @pilou- please review.

a. Replaces https://github.com/ansible/ansible/pull/25106 because @dkorn has less time now for it.
b. Takes care of reviews in https://github.com/ansible/ansible/pull/25106
c. Move to remote-managemet

Maintainers: @cben @dkorn @simon3z @yaacov

**Comment by yaacov** [comment]
cc @abraverm

**Comment by yaacov** [comment]
cc @moolitayer

**Comment by cben** [comment]
ready_for_review

**Comment by yaacov** [comment]
@thaumos please review

**Comment by mattclay** [comment]
There is a related PR https://github.com/ansible/ansible/pull/25865 which includes support for integration tests in CI.

**Comment by yaacov** [comment]
> There is a related PR #25865 which includes support for integration tests in CI.

@mattclay yes, we know about @abraverm 's work, we also discussed sharing parts of the code. Currently the two efforts affect too different parts of manageiq's life-cycle , #25865 deal with deployment and #26641 with management, we do not have enough in common to share code at this point.

**Comment by mattclay** [comment]
@yaacov Will you be able to leverage the integration test support in that PR for this one?

**Comment by yaacov** [comment]
> @yaacov Will you be able to leverage the integration test support in that PR for this one?

@mattclay #25865 's integration tests deal with the creation of new ManageIQ ( If I understand correctly, @abraverm please currect me if I'm wrong ) while #26641 assumes we have a working ManageIQ, I do not see a way to lavarage the integration tests from #25865, @cben, @abraverm can you see something I do not see ?

**Comment by mattclay** [comment]
@yaacov The links to those PRs appear to not be the ones you intended. Can you check?

**Comment by yaacov** [comment]
> @yaacov The links to those PRs appear to not be the ones you intended. Can you check?

Thanks 👍 edited the comment above, and fixed the links.

**Comment by mattclay** [comment]
@yaacov Since the tests in https://github.com/ansible/ansible/pull/25865 use a manageiq container, I would hope you'd be able to use the same container image to facilitate testing your PR as well. Possibly with some additional setup if the container isn't ready-to-use "out of the box".

**Comment by simon3z** [comment]
cc @gtanzillo

**Comment by yaacov** [comment]
@mattclay, thanks 👍 using #25865 for testing is great. Talked to @cben and @abraverm , it's very strait forward to use #25865 's testing infrastructure to facilitate testing for this PR, we prefer to add the tests after #25865 is merged, that way will not create duplications in code, and double work, WDYT ?

**Comment by yaacov** [comment]
@gtanzillo hi, how do we continue now ?

**Comment by gtanzillo** [comment]
@yaacov Need some time to review this. Will get to it in the next few days.

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (gtanzillo) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py:75]: It seems like the attributes of the user being created and the MIQ login info should be in separate lists. Is that possible?

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (gtanzillo) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: Perhaps this should be `userid` to match the MIQ side.  Also, `name` on the MIQ side is the user's actual full name and not sign on name.

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (gtanzillo) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: This one is `name` on the MIQ side.

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (gtanzillo) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: I think this should be a separate command option - `delete` or `create`. `state` seems like something different. At a minimum, I think it should be more explicit - like `action: create | delete`

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (gtanzillo) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: Like above, I think this should be `userid, name, password, group, email`

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (gtanzillo) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: How are the `user_details` used by the caller here? It looks like it's embedded in a string.

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (gtanzillo) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: Since `password` is not being passed to `user_update_required`, does that meant that updating the user's password is not supported?

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (gtanzillo) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: When `password` is not specified (this is a requirement for MIQ external auth) what get's passed to the MIQ API? Seem like it should test `password is not None` before building the json.

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_user.py:75]: the MIQ login moved into a separate list, named 'miq'

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: results are the return value of the `manageiq.client.post`, if the api call returns the new user object, it will returned as `user_details`

when running the example I get:
```
changed: [127.0.0.1] => {
    "changed": true, 
    "failed": false, 
    "invocation": {
        "module_args": {
            "action": "create", 
            "email": "jdoe@example.com", 
            "group": "EvmGroup-user", 
            "miq": {
                "password": "smartvm", 
                "url": "http://127.0.0.1:3000", 
                "username": "admin", 
                "verify_ssl": false
            }, 
            "name": "Jane Doe", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "userid": "jdoe"
        }
    }, 
    "msg": "Successfully created the user jdoe: [{u'name': u'Jane Doe', u'settings': {}, u'userid': u'jdoe', u'current_group_id': u'4', u'id': u'6', u'created_on': u'2017-07-26T09:54:45Z', u'href': u'http://127.0.0.1:3000/api/users/6', u'updated_on': u'2017-07-26T09:54:45Z', u'email': u'jdoe@example.com'}]"
}

```

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_user.py:75]: 👍 yes, moved to a list named 'miq'

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (cben) [reply lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: It's an API limitation we also had with [manageiq_provider](https://github.com/dkorn/manageiq-ansible-module/blob/master/library/manageiq_provider.py) module (future PR).
API (understandably) won't tell us what the passwords are now.  So you can update the password, as long as you update some other field :-(

- The best solution I see is always send the POST request.  This sounds expensive but isn't actually — we're always doing a GET now.  But we'd need API to tell us whether anything changed (ideally list of fields).  Returning truthful `changed` is crucial for ansible.

- Another solution is adding some `force_update_password` param.  Yikes.  IMHO it's better to wait and have partial functionality for now.

**Comment by yaacov** [comment]
@gtanzillo 👍 thanks for the review

I think I fixed the issues, please re-review

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: we try to use 'state' instead of action, as ti keeps with the declarative voice of Ansible vs an imperative approach from programming languages.

Comment (bcoca) [reply lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: keeping it absent vs present you guarantee how the user is supposed to be, the return changed state informs the user if there was an action taken or not, mostly users should not care if this is a creation or an edit, just that the result is what they want it to be.

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: no need to specify required=False or type=str as these are defaults

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: see comments above for 'state' vs 'action'

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: this is not good way to set 'changed', if user does not exist you are erroring out instead of returning 'ok' and `changed=False`

**Review by bcoca — CHANGES_REQUESTED** [review]
Most of the approach just mirrors the command with no thought of added value to the ansible user.

Look at the `user` module and how it handles the specific operations and lets the user only be concerned with desired state, not 'actions to get to that state'

**Comment by bcoca** [comment]
Before making any changes or new modules please read the following:

http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_best_practices.html
http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_checklist.html

everything in my review was already covered in these docs.

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: 👍 fixed

**Comment by yaacov** [comment]
@bcoca , changed the defaults and moved to using state, please re-review

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py:292]: You have the 'check_client' function to figure out if the requirements are present, but never actually use it, so the module fails on in import exception.

I recommend you place a call in the class init so you dont have to remember to call it in every module.

File: lib/ansible/utils/module_docs_fragments/manageiq.py
Comment (bcoca) [inline lib/ansible/utils/module_docs_fragments/manageiq.py]: so this is really an option with sub options, each entry should have it's own description/required/etc and the description for miq should be descriptive of what the grouping is.

**Review by bcoca — CHANGES_REQUESTED** [review]
previous issues fixed, just a couple of more things

File: lib/ansible/utils/module_docs_fragments/manageiq.py
Comment (yaacov) [reply lib/ansible/utils/module_docs_fragments/manageiq.py]: I tried, and did not find how to implement this :-(

When I wrote it as a regular `options:` list, rendered using `ansible-doc ` it was on the same level as the `managiq_user` `options:` and when I tried to set a new level of `['options']['options']` in the yaml, I got syntex errors form `ansible-doc `

What am I missing ? can you give a small example, how to do this ?

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_user.py:292]: 👍 done on the class init

File: lib/ansible/utils/module_docs_fragments/manageiq.py
Comment (yaacov) [reply lib/ansible/utils/module_docs_fragments/manageiq.py]: 👍 fixed in the doc fragment.

**Comment by yaacov** [comment]
@bcoca thanks 👍 , changed the documentation and the bad import handling, please re-review

**Comment by yaacov** [comment]
ready_for_review

**Comment by gtanzillo** [comment]
Gregg Tanzillo (@gtanzillo) and Alberto Bellotti (@abellotti) are the reviewers from ManageIQ.

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: format and a user provided field is a bit of a security issue `%` notation is better here, but not a showstopper

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: `{}` is not compatible across python versions that support format, use `{0}` or `%` instead

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: requiring this as a dict and not using suboptions nor no_log makes the connection password too exposed.

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: @can you refer me to a suboptions example, I am having trouble implementing it :-(

Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: @bcoca ^^ sorry I wrote the a wrong name

**Comment by yaacov** [comment]
@gtanzillo hi, I'm fixing the comments above from @bcoca . In the mean time, we need your ack ( or not ack :-) ) on the manageiq api stuff ...

cc @simon3z

**Review by gtanzillo — APPROVED** [review]
👍  LGTM

**Comment by yaacov** [comment]
@gtanzillo thanks , I will ping you again once I commit a fix to @bcoca 's "visible passwords on logs" issue

**Comment by yaacov** [comment]
@bcoca fixed the "visible password in logs" thing, please re-review

cc// @gtanzillo

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_user.py]: 👍 thanks, added a `no_log` here

**Comment by yaacov** [comment]
@bcoca @gundalow can I squash the individual commits (I left them separate for easy reviews) ?

**Comment by yaacov** [comment]
last commit is about using sub options, addressing a comment by @cben about (no_log) only the manageiq password and not all the connection object.

**Comment by yaacov** [comment]
@bcoca please re-review

**Comment by simon3z** [comment]
> Gregg Tanzillo (@gtanzillo) and Alberto Bellotti (@abellotti) are the reviewers from ManageIQ.

Zohar Galor (@zgalor) and Yaacov Zamir (@yaacov) are the reviewers from CM.

**Review by bcoca — APPROVED** [review]

**Pull request merged**
